// domain-digger/dist/test/traceroute-parser-test.js
"use strict";
var traceroute_parser_1 = require('../classes/parsers/traceroute-parser');
function parse(text) {
    var parser = new traceroute_parser_1.default(text);
    return parser.process();
}
function testValidRecord(test, rec, hopIndex, host, ipAddress, packetsTTLs) {
    test.strictEqual(rec.hopIndex, hopIndex, 'Hop indexes are not equal');
    test.strictEqual(rec.host, host, 'Host addresses are not equal');
    test.strictEqual(rec.ipAddress, ipAddress, 'IP addresses are not equal');
    for (var i = 0; i < packetsTTLs.length; ++i) {
        test.strictEqual(rec.packetsTTLs[i].time, packetsTTLs[i], 'TTLs are not equal');
    }
}
function testSilentRecord(test, rec, hopIndex, requiredPacketTTLs) {
    test.strictEqual(rec.hopIndex, hopIndex, 'Hop indexes are not equal');
    test.strictEqual(rec.host, null, 'Host is not empty');
    test.strictEqual(rec.ipAddress, null, 'IP address is not empty');
    if (requiredPacketTTLs) {
        for (var _i = 0, requiredPacketTTLs_1 = requiredPacketTTLs; _i < requiredPacketTTLs_1.length; _i++) {
            var ttl = requiredPacketTTLs_1[_i];
            var found = false;
            for (var _a = 0, _b = rec.packetsTTLs; _a < _b.length; _a++) {
                var packetTtl = _b[_a];
                if (packetTtl.time != ttl) {
                    continue;
                }
                found = true;
            }
            test.ok(found, "The time " + ttl + " not found in current records ttls");
        }
    }
}
function parseText1(test) {
    var records = parse(text1());
    test.equal(records.length, 64, 'Invalid records count');
    testValidRecord(test, records[0], 1, '192.168.200.1', '192.168.200.1', [3.013, 1.365, 1.757]);
    testSilentRecord(test, records[11], 12);
    test.done();
}
exports.parseText1 = parseText1;
function parseText2(test) {
    var records = parse(text2());
    test.equal(records.length, 15, 'Invalid records count');
    testValidRecord(test, records[3], 4, 'ge-4-1-1-1g.m120-1-klgd.nwtelecom.ru', '212.48.214.197', [1.922, 1.839, 1.882]);
    testValidRecord(test, records[9], 10, 'std-p2-be1.yndx.net', '87.250.239.133', [41.761, 47.590, 41.158]);
    testSilentRecord(test, records[6], 7);
    testSilentRecord(test, records[11], 12, [87.55]);
    test.done();
}
exports.parseText2 = parseText2;
function text1() {
    return "\n    traceroute to dnsrecords.com (192.64.74.229), 64 hops max, 52 byte packets\n 1  192.168.200.1 (192.168.200.1)  3.013 ms  1.365 ms  1.757 ms\n 2  c7206-11-klgd.baltnet.ru (217.168.64.70)  2.732 ms  1.616 ms  1.507 ms\n 3  erx1440-1-klgd.baltnet.ru (217.168.64.69)  2.619 ms  2.303 ms  2.085 ms\n 4  ge-4-1-1-1g.m120-1-klgd.nwtelecom.ru (212.48.214.197)  1.830 ms  2.084 ms  1.871 ms\n 5  188.254.78.214 (188.254.78.214)  1.986 ms  1.938 ms  1.816 ms\n 6  87.226.133.147 (87.226.133.147)  30.717 ms  30.457 ms  30.587 ms\n 7  ae23-xcr1.fri.cw.net (213.185.219.105)  47.240 ms  47.436 ms  47.063 ms\n 8  ae14-xcr1.ptl.cw.net (195.2.31.234)  130.131 ms  131.300 ms  130.170 ms\n 9  et-7-1-0-xcr1.nyh.cw.net (195.2.24.241)  127.166 ms  126.970 ms  127.152 ms\n10  ae13-xcr2.nyk.cw.net (195.2.25.69)  129.551 ms  129.818 ms  129.607 ms\n11  nyc2-brdr-01.inet.qwest.net (63.146.27.61)  127.152 ms  127.626 ms  129.563 ms\n12  * * *\n13  63-235-80-54.dia.static.qwest.net (63.235.80.54)  193.279 ms  192.775 ms  190.430 ms\n14  border1.t7-1-bbnet1.sef003.pnap.net (63.251.160.20)  190.805 ms  190.825 ms  193.063 ms\n15  bdhg10g-7.border1.sef003.pnap.net (63.251.175.250)  199.013 ms  198.943 ms  199.046 ms\n16  * * *\n17  * * *\n18  * * *\n19  * * *\n20  * * *\n21  * * *\n22  * * *\n23  * * *\n24  * * *\n25  * * *\n26  * * *\n27  * * *\n28  * * *\n29  * * *\n30  * * *\n31  * * *\n32  * * *\n33  * * *\n34  * * *\n35  * * *\n36  * * *\n37  * * *\n38  * * *\n39  * * *\n40  * * *\n41  * * *\n42  * * *\n43  * * *\n44  * * *\n45  * * *\n46  * * *\n47  * * *\n48  * * *\n49  * * *\n50  * * *\n51  * * *\n52  * * *\n53  * * *\n54  * * *\n55  * * *\n56  * * *\n57  * * *\n58  * * *\n59  * * *\n60  * * *\n61  * * *\n62  * * *\n63  * * *\n64  * * *\n    ";
}
function text2() {
    return "\ntraceroute: Warning: ya.ru has multiple addresses; using 93.158.134.3\ntraceroute to ya.ru (93.158.134.3), 15 hops max, 52 byte packets\n 1  192.168.200.1 (192.168.200.1)  1.384 ms  1.769 ms  1.365 ms\n 2  c7206-11-klgd.baltnet.ru (217.168.64.70)  2.542 ms  1.800 ms  1.572 ms\n 3  erx1440-1-klgd.baltnet.ru (217.168.64.69)  2.222 ms  2.325 ms  2.212 ms\n 4  ge-4-1-1-1g.m120-1-klgd.nwtelecom.ru (212.48.214.197)  1.922 ms  1.839 ms  1.882 ms\n 5  212.48.195.155 (212.48.195.155)  32.150 ms  32.259 ms  32.099 ms\n 6  ppp-166.as001.spbnit.ru (212.48.197.166)  32.379 ms  32.449 ms  32.359 ms\n 7  * * *\n 8  ppp-189.as001.spbnit.ru (212.48.197.189)  33.274 ms  32.666 ms  33.854 ms\n 9  msk-ix-std.yandex.net (195.208.208.116)  44.857 ms  42.773 ms  41.937 ms\n10  std-p2-be1.yndx.net (87.250.239.133)  41.761 ms\n    m9-p2-eth-trunk8.yndx.net (87.250.239.127)  47.590 ms\n    std-p2-be1.yndx.net (87.250.239.133)  41.158 ms\n11  ugr-b-c1-ae5.yndx.net (87.250.239.53)  43.319 ms  41.188 ms\n    ugr-b-c1-ae13.yndx.net (87.250.239.86)  40.865 ms\n12  * 87.55 *\n13  www.yandex.ru (93.158.134.3)  48.006 ms *  53.366 ms\n14  * * *\n15  www.yandex.ru (93.158.134.3)  45.897 ms  47.515 ms  41.532 ms\n    ";
}
